File: engine/src/main/java/org/epdb/engine/volcano/Insert.java
package org.epdb.engine.volcano;

import org.epdb.buffer.BufferManager;
import org.epdb.engine.columntypes.ColumnValue;
import org.epdb.engine.columntypes.IntValue;
import org.epdb.engine.columntypes.StringValue;
import org.epdb.engine.dto.Tuple;
import org.epdb.index.IndexManager;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.StandardCharsets;

import static org.epdb.storage.dto.Page.*;

public class Insert implements Operator {

    private static final int NAME_SIZE = 20;
    private static final int INDEXED_COLUMN_ID = 0;

    private final IndexManager indexManager;
    private final BufferManager bufferManager;
    private final Tuple tupleToInsert;
    private final Long tableStartPageId;
    private final int maxAllocatedPageCount;

    private boolean isExecuted;

    public Insert(final BufferManager bufferManager, final Tuple tupleToInsert, final int maxAllocatedPageCount, Long tableStartPageId, final IndexManager indexManager) {
        this.indexManager = indexManager;
        this.bufferManager = bufferManager;
        this.tupleToInsert = tupleToInsert;
        this.isExecuted = false;
        this.tableStartPageId = tableStartPageId;
        this.maxAllocatedPageCount = maxAllocatedPageCount;
    }

    @Override
    public void open() {}

    @Override
    public Tuple next() {
        if(this.isExecuted) {
            return null;
        }

        this.isExecuted = true;
        var serializedTuple = serializeTuple(this.tupleToInsert);

        for(var currentPageId = this.tableStartPageId; currentPageId <= this.maxAllocatedPageCount; currentPageId++) {
            var page = bufferManager.getPage(currentPageId);

            var currentFreeSpaceOffset = page.getFreeSpaceOffset();
            var currentNumSlots = page.getCurrentNumSlots();
            var nextSlotAddress = PAGE_SIZE_IN_BYTES - ((currentNumSlots + 1) * SLOT_SIZE_IN_BYTES);

            if (currentFreeSpaceOffset + serializedTuple.length < nextSlotAddress) {
                updateColumnIndex(tupleToInsert.getValueAtIndex(INDEXED_COLUMN_ID), currentNumSlots, page.getPageId());
                page.writeTupleAndSlot(serializedTuple);
                System.out.println("Inserted tuple: " + tupleToInsert);
                bufferManager.unpinPage(currentPageId, true);
                return this.tupleToInsert;
            } else {
                bufferManager.unpinPage(currentPageId, false);
            }
        }

        var page = this.bufferManager.allocateNewPage(0);
        page.writeTupleAndSlot(serializedTuple);
        updateColumnIndex(tupleToInsert.getValueAtIndex(INDEXED_COLUMN_ID), 0, page.getPageId());;
        System.out.println("Inserted tuple: " + tupleToInsert);
        return this.tupleToInsert;
    }

    @Override
    public void close() {}

    private void updateColumnIndex(ColumnValue key, int slotIndex, Long pageId) {
        indexManager.addEntry(key, pageId, slotIndex);
    }

    private byte[] serializeTuple(Tuple tuple) {
        var id = (IntValue) tuple.getValueAtIndex(0);
        var name = (StringValue) tuple.getValueAtIndex(1);
        var nameBytes = name.getValue().getBytes(StandardCharsets.UTF_8);
        var age = (IntValue) tuple.getValueAtIndex(2);
        var byteBuffer = ByteBuffer.allocate(ROW_SIZE_IN_BYTES).order(ByteOrder.LITTLE_ENDIAN);

        byteBuffer.putInt(id.getValue());
        byteBuffer.put(nameBytes, 0, Math.min(nameBytes.length, NAME_SIZE));
        for (int i = nameBytes.length; i < NAME_SIZE; i++) {
            byteBuffer.put((byte) 0);
        }
        byteBuffer.putInt(age.getValue());

        return byteBuffer.array();
    }
}

File: modules/engine/src/main/kotlin/org/epdb/engine/databaseoperator/Insert.kt
package org.epdb.engine.databaseoperator

import org.epdb.buffer.manager.BufferManager
import org.epdb.engine.columntypes.ColumnValue
import org.epdb.engine.columntypes.IntValue
import org.epdb.engine.columntypes.StringValue
import org.epdb.engine.dto.Tuple
import org.epdb.index.manager.IndexManager
import org.epdb.storage.dto.Page
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.charset.StandardCharsets
import kotlin.math.min

class Insert(
    private val bufferManager: BufferManager,
    private val tupleToInsert: Tuple,
    private val maxAllocatedPageCount: Long,
    private val tableStartPageId: Long,
    private val indexManager: IndexManager,
    private var isExecuted: Boolean = false
) : Operator {

    companion object {
        private const val NAME_SIZE = 20
        private const val INDEXED_COLUMN_ID = 0
    }

    override fun open() {}

    override fun next(): Tuple? {
        if (this.isExecuted) return null
        this.isExecuted = true

        val serializedTuple = serializeTuple(tupleToInsert)

        for (currentPageId in tableStartPageId..maxAllocatedPageCount) {
            val page = bufferManager.getPage(currentPageId)

            try {
                if (canInsertTuple(page, serializedTuple.size)) {
                    updateColumnIndex(tupleToInsert.getValueAtIndex(INDEXED_COLUMN_ID), page.currentNumSlots, page.pageId)
                    page.writeTupleAndSlot(serializedTuple)
                    println("Inserted tuple: $tupleToInsert")
                    bufferManager.unpinPage(currentPageId, true)
                    return this.tupleToInsert
                }

                bufferManager.unpinPage(currentPageId, false)
            } catch (e: Exception) {
                bufferManager.unpinPage(currentPageId, false)
                throw e
            }
        }

        val newPage = this.bufferManager.allocateNewPage(0)
        try {
            updateColumnIndex(tupleToInsert.getValueAtIndex(INDEXED_COLUMN_ID), 0, newPage.pageId)
            newPage.writeTupleAndSlot(serializedTuple)
            println("Inserted tuple: $tupleToInsert")
            return this.tupleToInsert
        } finally {
            bufferManager.unpinPage(newPage.pageId, true)
        }
    }

    override fun close() {}

    private fun canInsertTuple(page: Page, requiredSize: Int) : Boolean {
        val currentNumSlots = page.currentNumSlots
        val nextSlotAddress = Page.PAGE_SIZE_IN_BYTES - ((currentNumSlots + 1) * Page.SLOT_SIZE_IN_BYTES)

        return page.freeSpaceOffset + requiredSize < nextSlotAddress
    }

    private fun updateColumnIndex(key: ColumnValue, slotIndex: Int, pageId: Long) {
        indexManager.addEntry(key, pageId, slotIndex)
    }

    private fun serializeTuple(tuple: Tuple): ByteArray {
        val id = tuple.getValueAtIndex(0) as? IntValue ?: throw IllegalArgumentException("Missing or invalid ID")
        val name = tuple.getValueAtIndex(1) as? StringValue ?: throw IllegalArgumentException("Missing or invalid Name")
        val age = tuple.getValueAtIndex(2) as? IntValue ?: throw IllegalArgumentException("Missing or invalid Age")

        val nameBytes = name.value.toByteArray(StandardCharsets.UTF_8)
        val byteBuffer = ByteBuffer.allocate(Page.ROW_SIZE_IN_BYTES).order(ByteOrder.LITTLE_ENDIAN)

        byteBuffer.putInt(id.value)
        byteBuffer.put(nameBytes, 0, min(nameBytes.size, NAME_SIZE))
        for (i in nameBytes.size..<NAME_SIZE) {
            byteBuffer.put(0.toByte())
        }
        byteBuffer.putInt(age.value)

        return byteBuffer.array()
    }
}
